### ISSUE TITLE
URLPattern `{/:optional}?` no longer works as expected.

### ISSUE THREAD

**Issue by JTRNS** [denoland/fresh#2693]
**TLDR**

```ts
const pattern = new URLPattern({pathname: "/books{/:genre}?" });
console.log(pattern.pathname) // PREVIOUSLY resulted in "/books{/:genre}?"
console.log(pattern.pathname) // CURRENTLY results in "/books/:genre?"
```

Just wanted to give you a heads up on some additional `URLPattern` changes that impact [the current router logic inside the `pathToPattern` function](https://github.com/denoland/fresh/blob/e303b873a6b66a5cbba7de152d66e029520ba21d/src/router.ts#L193C17-L199C21).

```ts
// snippet from /src/router.ts
if (part[j + 1] === "[") {
  // Disallow optional dynamic params like `foo-[[bar]]`
  if (part[j - 1] !== "/" && !!part[j - 1]) {
    throw new SyntaxError(
      `Invalid route pattern: "${path}". An optional parameter needs to be a full segment.`,
    );
  }
  groupOpen++;
  optional = true;
  pattern += "{/";
  j++;
}
```

Haven't exactly been able to figure out why, but I did notice that [the test case for this pattern](https://github.com/denoland/deno/pull/25961/commits/db873f0a9b3f6d4111e4938bf875b4fb6459eb39#diff-ef0c027d3b7fae135e81335bbb595c66f7d3854671829a98ffbafbdd9dca6f51) was recently removed from the deno repository.

I am assuming it simply didn't comply with the specification, but not completely sure. If you have some more insight into the _why_ this URL pattern is no longer supported I'd love to know as I was using it too.

**Comment by marvinhagemeister** [comment]
There was a recent specification change in `URLPattern`, see https://github.com/whatwg/urlpattern/pull/198 . It allows group values to be `undefined` instead of always falling back to an empty `string`.

Taking your pattern as an example: The `genre` group can now be `undefined` when passing `http://example.com/books` without a genre.

```ts
// Before
> new URLPattern({pathname: "/books{/:genre}?" }).exec("https://foo.com/books")
{
  inputs: [ "https://foo.com/books" ],
  protocol: { input: "https", groups: { "0": "https" } },
  username: { input: "", groups: { "0": "" } },
  password: { input: "", groups: { "0": "" } },
  hostname: { input: "foo.com", groups: { "0": "foo.com" } },
  port: { input: "", groups: { "0": "" } },
  pathname: { input: "/books", groups: { genre: "" } },   //  <-- falls back to empty string
  search: { input: "", groups: { "0": "" } },
  hash: { input: "", groups: { "0": "" } }
}

// After
> new URLPattern({pathname: "/books{/:genre}?" }).exec("https://foo.com/books")
{
  inputs: [ "https://foo.com/books" ],
  protocol: { input: "https", groups: { "0": "https" } },
  username: { input: "", groups: { "0": "" } },
  password: { input: "", groups: { "0": "" } },
  hostname: { input: "foo.com", groups: { "0": "foo.com" } },
  port: { input: "", groups: { "0": "" } },
  pathname: { input: "/books", groups: { genre: undefined } },   //  <-- is undefined
  search: { input: "", groups: { "0": "" } },
  hash: { input: "", groups: { "0": "" } }
}
```

In Fresh we always normalize it to a `string` for now to avoid introducing a breaking change to users.
